expression/errors.go

```diff
@@ -0,0 +1,60 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expression
+
+import (
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/terror"
+	"github.com/pingcap/tidb/util/types"
+)
+
+// Error instances.
+var (
+	ErrIncorrectParameterCount = terror.ClassExpression.New(mysql.ErrWrongParamcountToNativeFct, mysql.MySQLErrName[mysql.ErrWrongParamcountToNativeFct])
+
+	errInvalidOperation    = terror.ClassExpression.New(codeInvalidOperation, "invalid operation")
+	errFunctionNotExists   = terror.ClassExpression.New(mysql.ErrSpDoesNotExist, mysql.MySQLErrName[mysql.ErrSpDoesNotExist])
+	errZlibZData           = terror.ClassTypes.New(mysql.ErrZlibZData, mysql.MySQLErrName[mysql.ErrZlibZData])
+	errIncorrectArgs       = terror.ClassExpression.New(mysql.ErrWrongArguments, mysql.MySQLErrName[mysql.ErrWrongArguments])
+	errUnknownCharacterSet = terror.ClassExpression.New(mysql.ErrUnknownCharacterSet, mysql.MySQLErrName[mysql.ErrUnknownCharacterSet])
+)
+
+// Error codes.
+const (
+	codeInvalidOperation terror.ErrCode = 1
+)
+
+func init() {
+	expressionMySQLErrCodes := map[terror.ErrCode]uint16{
+		mysql.ErrWrongParamcountToNativeFct: mysql.ErrWrongParamcountToNativeFct,
+		mysql.ErrSpDoesNotExist:             mysql.ErrSpDoesNotExist,
+		mysql.ErrZlibZData:                  mysql.ErrZlibZData,
+		mysql.ErrWrongArguments:             mysql.ErrWrongArguments,
+	}
+	terror.ErrClassToMySQLCodes[terror.ClassExpression] = expressionMySQLErrCodes
+}
+
+// handleInvalidTimeError reports error or warning depend on the context.
+func handleInvalidTimeError(ctx context.Context, err error) error {
+	if err == nil || !(terror.ErrorEqual(err, types.ErrInvalidTimeFormat) || types.ErrIncorrectDatetimeValue.Equal(err)) {
+		return err
```

Comment:
Add the error trace?

Reply:
no, this function should never touch the input error, it just decides whether to append the input error to warnings.